From b42661a086c8e0bd7eff1c1fc1606ab92b489322 Mon Sep 17 00:00:00 2001
From: Sandy Huang <hjc@rock-chips.com>
Date: Tue, 6 Mar 2018 09:25:12 +0800
Subject: [PATCH] drm/rockchip: framebuffer should never be freed

In recovery mode, fb will be freed and lead to following error:

[    2.571563] ffffffc03d5f9400: FB ID: 0 (0) fb:0xffffffc03d5f9400, fb->refcount:0
[    2.571694] ------------[ cut here ]------------
[    2.571717] WARNING: at include/linux/kref.h:46
[    2.571738] Modules linked in:
[    2.571763]
[    2.571792] CPU: 0 PID: 169 Comm: recovery Not tainted 4.4.114 #1042
[    2.571815] Hardware name: Rockchip rk3326 863 board (DT)
[    2.571843] task: ffffffc03b078d40 task.stack: ffffffc03b064000
[    2.571877] PC is at drm_framebuffer_reference+0x98/0xc8
[    2.571910] LR is at drm_framebuffer_reference+0x68/0xc8
......
[    2.589269] [<ffffff80084823a8>] drm_framebuffer_reference+0x98/0xc8
[    2.589306] [<ffffff80084831ac>] drm_mode_set_config_internal+0xd4/0x110
[    2.589339] [<ffffff8008488560>] drm_mode_setcrtc+0x448/0x4ec
[    2.589370] [<ffffff8008479830>] drm_ioctl+0x26c/0x3f4
[    2.589404] [<ffffff80081d3884>] do_vfs_ioctl+0x6f0/0x82c
[    2.589438] [<ffffff80081d3a20>] SyS_ioctl+0x60/0x88
[    2.589470] [<ffffff80080832f0>] el0_svc_naked+0x24/0x28

Change-Id: I0706c832b4705bf23147c306e34557a152fb069b
Signed-off-by: Sandy Huang <hjc@rock-chips.com>
---
 drivers/gpu/drm/rockchip/rockchip_drm_drv.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/gpu/drm/rockchip/rockchip_drm_drv.c b/drivers/gpu/drm/rockchip/rockchip_drm_drv.c
index 959ba8297845..752bdb111b05 100644
--- a/drivers/gpu/drm/rockchip/rockchip_drm_drv.c
+++ b/drivers/gpu/drm/rockchip/rockchip_drm_drv.c
@@ -1404,6 +1404,7 @@ static int rockchip_drm_bind(struct device *dev)
 	drm_for_each_crtc(crtc, drm_dev) {
 		struct drm_fb_helper *helper = private->fbdev_helper;
 		crtc->primary->fb = helper->fb;
+		drm_framebuffer_reference(helper->fb);
 	}
 	drm_dev->mode_config.allow_fb_modifiers = true;
 
-- 
2.35.3

